swift/xcode-autobuilder/xcode-autobuilder.cpp

```diff
@@ -44,38 +38,63 @@ static bool endsWith(std::string_view s, std::string_view suffix) {
   return t.fileCount == 0 || t.type == uiTest || t.type == unitTest ||
          // unknown target types can be legitimate, let's do a name-based heuristic then
          (t.type == unknownType && (endsWith(t.name, "Tests") || endsWith(t.name, "Test")));
 }
 
-static void autobuild(const CLIArgs& args) {
-  auto collected = collectTargets(args.workingDir);
-  auto& targets = collected.targets;
-  for (const auto& t : targets) {
+static void buildSwiftPackages(const std::vector<std::filesystem::path>& swiftPackages,
+                               bool dryRun) {
+  auto any_successful =
+      std::any_of(std::begin(swiftPackages), std::end(swiftPackages), [&](auto& packageFile) {
+        LOG_INFO("Building Swift package: {}", packageFile);
+        return buildSwiftPackage(packageFile, dryRun);
+      });
+  if (!any_successful) {
+    codeql::Log::flush();
+    exit(1);
+  }
+}
+
+static bool autobuild(const CLIArgs& args) {
+  auto structure = scanProjectStructure(args.workingDir);
+  auto& xcodeTargets = structure.xcodeTargets;
+  auto& swiftPackages = structure.swiftPackages;
+  for (const auto& t : xcodeTargets) {
     LOG_INFO("{}", t);
   }
   // Filter out targets that are tests or have no swift source files
-  targets.erase(std::remove_if(std::begin(targets), std::end(targets), isNonSwiftOrTestTarget),
-                std::end(targets));
+  xcodeTargets.erase(
+      std::remove_if(std::begin(xcodeTargets), std::end(xcodeTargets), isNonSwiftOrTestTarget),
+      std::end(xcodeTargets));
 
   // Sort targets by the amount of files in each
-  std::sort(std::begin(targets), std::end(targets),
-            [](Target& lhs, Target& rhs) { return lhs.fileCount > rhs.fileCount; });
-  if ((!collected.xcodeEncountered || targets.empty()) && collected.swiftPackageEncountered) {
-    DIAGNOSE_ERROR(spmNotSupported,
-                   "A Swift package was detected, but no viable Xcode target was found.");
-  } else if (!collected.xcodeEncountered) {
-    DIAGNOSE_ERROR(noProjectFound, "`autobuild` could not detect an Xcode project or workspace.");
-  } else if (targets.empty()) {
+  std::sort(std::begin(xcodeTargets), std::end(xcodeTargets),
+            [](XcodeTarget& lhs, XcodeTarget& rhs) { return lhs.fileCount > rhs.fileCount; });
+
+  if (structure.xcodeEncountered && xcodeTargets.empty() && swiftPackages.empty()) {
+    // Report error only when there are no Xcode targets and no Swift packages
     DIAGNOSE_ERROR(noSwiftTarget, "All targets found within Xcode projects or workspaces either "
                                   "contain no Swift source files, or are tests.");
-  } else {
-    LOG_INFO("Selected {}", targets.front());
-    buildTarget(targets.front(), args.dryRun);
-    return;
+    return false;
+  } else if (!structure.xcodeEncountered && swiftPackages.empty()) {
+    DIAGNOSE_ERROR(noProjectFound,
+                   "`autobuild` could not detect an Xcode project or workspace or Swift package");
```

Comment:
wording here is a bit ambiguous grammatically (you mean `(Xcode project or workspace) or Swift package`, but it could theoretically mean `Xcode (project or workspace or Swift package)`. So maybe
```suggestion
                   "`autobuild` detected neither an Xcode project or workspace, nor a Swift package");
```